lmdeploy/pytorch/disagg/backend/mooncake.py

```diff
@@ -1,28 +1,256 @@
 # Copyright (c) OpenMMLab. All rights reserved.
+
+import json
+import os
+import socket
+import subprocess
+from typing import Dict
+
 from lmdeploy.pytorch.disagg.backend.backend import MIGRATION_BACKENDS
 from lmdeploy.pytorch.disagg.backend.base import MigrationBackendImpl
-from lmdeploy.pytorch.disagg.config import MigrationBackend, MigrationProtocol
+from lmdeploy.pytorch.disagg.config import DistServeEngineConfig, MigrationBackend, MigrationProtocol
 from lmdeploy.pytorch.disagg.messages import DistServeRegisterMRMessage, MigrationAssignment
 from lmdeploy.pytorch.disagg.request import DistServeConnectionRequest, DistServeInitRequest
+from lmdeploy.utils import get_logger
+
+logger = get_logger('lmdeploy')
+
+LMDEPLOY_USE_ASYNC_MIGRATION = os.environ.get('LMDEPLOY_USE_ASYNC_MIGRATION', None)
+
+
+def get_rdma_nics():
+    """Get all available RDMA network interface cards on the current machine.
+
+    Returns:
+        list: List of RDMA NICs, e.g. ['erdma_0', 'erdma_1']
+    """
+    rdma_nics = []
+
+    try:
+        result = subprocess.run(['ibv_devices'], stdout=subprocess.PIPE, text=True)
+        if result.returncode == 0:
+            # Parse ibv_devices output
+            # Sample output:
+            # device                 node GUID
+            # ------              ----------------
+            # erdma_0             02163efffe3fc264
+            # erdma_1             02163efffe3fc317
+            lines = result.stdout.strip().split('\n')
+            for line in lines[2:]:  # Skip header lines
+                if line.strip():
+                    device_name = line.split()[0].strip()
+                    rdma_nics.append(device_name)
+    except Exception as e:
+        print(f'Error executing ibv_devices command: {e}')
+
+    return rdma_nics
+
+
+def get_local_ip_by_remote() -> str:
+    # try ipv4
+    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+    try:
+        s.connect(('8.8.8.8', 80))  # Doesn't need to be reachable
+        return s.getsockname()[0]
+    except Exception:
+        pass
+
+    # try ipv6
+    try:
+        s = socket.socket(socket.AF_INET6, socket.SOCK_DGRAM)
+        # Google's public DNS server, see
+        # https://developers.google.com/speed/public-dns/docs/using#addresses
+        s.connect(('2001:4860:4860::8888', 80))  # Doesn't need to be reachable
+        return s.getsockname()[0]
+    except Exception:
+        raise ValueError('Can not get local ip')
+
+
+class MooncakeMigrationManagement:
+    """Manages migration for a single connection in Mooncake backend."""
+
+    def __init__(self, init_request: DistServeInitRequest):
+        try:
+            from mooncake.engine import TransferEngine
+        except ImportError as e:
+            raise ImportError('Please install mooncake by following the instructions at '
```

Comment:
When passing --migration-backend Mooncake, it's better to raise an import error immediately if Mooncake is not installed during API server launch.
Can we put it in the constructor of MooncakeBackend?